rfc/0022-policy-lifecycle.md

```diff
@@ -0,0 +1,397 @@
+
+#### Limitations: Compatibility of Precompiled Modules
+
+Precompiled modules can only be reused by another `wasmtime::Engine` if the following conditions are met:
+
+1. **Same Wasmtime version**: Both the "optimizer" (the engine that precompiled the module) and the end consumer must use the same version of Wasmtime.
+2. **Same Wasmtime configuration**: Both must use the same `wasmtime::Config`. For example, enabling features like [async support](https://docs.rs/wasmtime/latest/wasmtime/struct.Config.html#method.async_support) or [epoch interruption](https://docs.rs/wasmtime/latest/wasmtime/struct.Config.html#method.epoch_interruption) on the consumer side, while the optimizer does not, will result in a loading error.
+
+This is documented [here](https://docs.rs/wasmtime/latest/wasmtime/struct.Module.html#method.deserialize):
+
+> Note that this function is designed to be safe receiving output from any compiled version of wasmtime itself.
+> This means that it is safe to feed output from older versions of Wasmtime into this function, in addition to newer versions of wasmtime (from the future!).
+> These inputs will deterministically and safely produce an Err. This function only successfully accepts inputs from the same version of wasmtime,
+> but the safety guarantee only applies to externally-defined blobs of bytes, not those defined by any version of wasmtime.
+> (this means that if you cache blobs across versions of wasmtime you can be safely guaranteed that future versions of wasmtime will reject old cache entries).
+
+Wasmtime ensures safety by producing an error when an incompatible version of a module is loaded. This allows us to handle such errors gracefully by downloading and re-optimizing the module as needed.
+Alternatively, we could save metadata alongside the binary blob containing the version and configuration details. Wasmtime includes mechanisms for version tracking to assist in ensuring compatibility.
+Please refer to the [Wasmtime documentation](https://docs.rs/wasmtime/latest/wasmtime/struct.Config.html#method.module_version) for more information about module versions strategies.
+
+Therefore, we need to store precompiled modules in a path that includes the Policy Server version.
+This ensures compatibility when the Policy Server is updated, and the precompiled modules might no longer work with the new version.
+Example: `/cache/ghcr.io/kubewarden/policies/safe-labels/v1.0.0/optimized-kw-v1.28.1.wasm`.
+TODO: explain this example
+
+#### Security Implications
+
+Loading precompiled binary blobs from a shared cache introduces potential security risks.
+An attacker could replace a valid binary blob with one that runs arbitrary malicious code.
+
+To mitigate this risk, we will use [sigstore](https://github.com/sigstore)'s signing capabilities to sign and verify all binary blobs.
```

Comment:
We should ensure this is possible offline to be compatible with airgap. Cosign has offline verification, but needs network for signing (see [here](https://github.com/sigstore/cosign/blob/main/README.md#verify-a-container-in-an-air-gapped-environment)).